packages/core/src/indexing/service.ts

```diff
@@ -1,1187 +1,724 @@
-        state.tasksLoadedFromCheckpoint = tasks[0].data.checkpoint;
-      }
+export const kill = (indexingService: IndexingService) => {
+  indexingService.common.logger.debug({
+    service: "indexing",
+    msg: "Killed indexing service",
+  });
+  indexingService.isKilled = true;
+};
 
-      // Add a flag to emit logs and checkpoints
-      tasks[tasks.length - 1].data.eventsProcessed = events.length;
+export const updateLastEventCheckpoint = (
+  indexingService: IndexingService,
+  lastEventCheckpoint: Checkpoint,
+) => {
+  indexingService.lastEventCheckpoint = lastEventCheckpoint;
 
-      // Update firstEventCheckpoint
-      if (state.firstEventCheckpoint === undefined) {
-        state.firstEventCheckpoint = tasks[0].data.checkpoint;
-      }
-    } else {
-      state.tasksLoadedFromCheckpoint = state.tasksLoadedToCheckpoint;
+  if (indexingService.firstEventCheckpoint !== undefined) {
+    indexingService.common.metrics.ponder_indexing_total_seconds.set(
+      indexingService.lastEventCheckpoint.blockTimestamp -
+        indexingService.firstEventCheckpoint.blockTimestamp,
+    );
+  }
+};
 
-      this.emitCheckpoint();
-    }
+const executeSetup = async (
+  indexingService: IndexingService,
+  { event }: { event: SetupEvent },
+): Promise<{ status: "error"; error: Error } | { status: "success" }> => {
+  const {
+    common,
+    indexingFunctions,
+    currentEvent,
+    networkByChainId,
+    contractsByChainId,
+    clientByChainId,
+  } = indexingService;
+  const indexingFunction =
+    indexingFunctions[event.contractName][event.eventName];
+
+  const metricLabel = {
+    event: event.eventName,
+    network: networkByChainId[event.chainId].name,
+  };
 
-    // Update lastEventCheckpoint
-    if (
-      lastCheckpoint === undefined &&
-      state.lastEventCheckpoint === undefined
-    ) {
-      // Fully cached path, first load
-      state.lastEventCheckpoint = toCheckpoint;
-
-      this.logCachedProgress(key);
-      this.updateTotalSeconds(key);
-      this.updateCompletedSeconds(key);
-    } else if (
-      lastCheckpoint !== undefined &&
-      state.lastEventCheckpoint === undefined
-    ) {
-      // Partially cached path, first load
-      state.lastEventCheckpoint = lastCheckpoint;
-
-      this.logCachedProgress(key);
-      this.updateTotalSeconds(key);
-      this.updateCompletedSeconds(key);
-    } else if (
-      lastCheckpoint !== undefined &&
-      state.lastEventCheckpoint !== undefined
-    ) {
-      // Subsequent loads
-      state.lastEventCheckpoint = checkpointMax(
-        lastCheckpoint,
-        state.lastEventCheckpoint,
+  for (let i = 0; i < 4; i++) {
+    try {
+      // set currentEvent
+      currentEvent.context.network.chainId = event.chainId;
+      currentEvent.context.network.name = networkByChainId[event.chainId].name;
+      currentEvent.context.client = clientByChainId[event.chainId];
+      currentEvent.context.contracts = contractsByChainId[event.chainId];
```

Comment:
Based on this, it looks like you could construct a `contextByChainId` mapping to encapsulate all of this, and then only update the context state here.